### PR TITLE
fixed issue #67 and #51

### LINKED ISSUE #51: Change `tables` to use :potential instead of :p
The DataFrames used as factors have a `:p` column. P is fairly likely to be a variable name.
It might be better to use `:potential` instead.

Thoughts?

### LINKED ISSUE #67: Use package-local wrapper type instead of extending Base (or other packages') methods on DataFrame
Since https://github.com/sisl/BayesNets.jl/blob/5febf756d38194712cfb8c0c2294252931611afa/src/DiscreteBayesNet/tables.jl does a `const Table = DataFrame`, all method extensions on `Table` are global and change the behavior of other unrelated code that uses DataFrames, just from importing BayesNets whether or not it's using this package. This is [discouraged](https://docs.julialang.org/en/latest/manual/style-guide.html#Avoid-type-piracy-1).

If `Table` were instead a wrapper struct that delegated the relevant operations to a single DataFrame field, then you could define methods on `BayesNets.Table` however you like and not be at risk of changing any other code's behavior.

Same issue with https://github.com/sisl/BayesNets.jl/blob/5febf756d38194712cfb8c0c2294252931611afa/src/ProbabilisticGraphicalModels/assignments.jl#L7 and https://github.com/sisl/BayesNets.jl/blob/5febf756d38194712cfb8c0c2294252931611afa/src/ProbabilisticGraphicalModels/nodenames.jl#L5 as well

and https://github.com/sisl/BayesNets.jl/commit/3cd56dc6a99e0a650177577dda3821b1fd7a0244#commitcomment-22494521 should probably be submitted upstream in DataFrames?

### PR DESCRIPTION
**Pull request by dwijenchawra**
fixed #67, fixed #51

cannot pass builds as there is an issue with lualatex dependency downloading, something to do with ubuntu's package manager not getting the right download link for imagemagick

### CODE REVIEW

**Comment by dwijenchawra** [comment]
added tests as well for documentation building

File: src/CPDs/categorical_cpd.jl [hunk 1]
```diff
@@ -37,7 +37,7 @@ function (cpd::CategoricalCPD)(a::Assignment=Assignment())
     if isempty(cpd.parents)
         return first(cpd.distributions)
     else
-        sub = [a[p] for p in cpd.parents]
+        sub = [a[potential] for potential in cpd.parents]
```
Comment (mykelk) [inline src/CPDs/categorical_cpd.jl:40]: the original `p` should stay. It is referring to the parents as `p`, not potentials.

File: src/CPDs/categorical_cpd.jl [hunk 2]
```diff
@@ -84,7 +84,7 @@ function Distributions.fit(::Type{CategoricalCPD{D}},
     # calc parent_instantiation_counts
 
     nparents = length(parents)
-    parental_ncategories = map!(p->infer_number_of_instantiations(data[!,p]), Array{Int}(undef, length(parents)), parents)
+    parental_ncategories = map!(potential->infer_number_of_instantiations(data[!,potential]), Array{Int}(undef, length(parents)), parents)
```
Comment (mykelk) [inline src/CPDs/categorical_cpd.jl:87]: should be `p` for parents, not potential

File: src/CPDs/categorical_cpd.jl [hunk 3]
```diff
@@ -119,7 +119,7 @@ function Distributions.fit(::Type{DiscreteCPD},
     data::DataFrame,
     target::NodeName,
     parents::NodeNames;
-    parental_ncategories::Vector{Int} = map!(p->infer_number_of_instantiations(data[!,p]), Array{Int}(undef, length(parents)), parents),
+    parental_ncategories::Vector{Int} = map!(potential->infer_number_of_instantiations(data[!,potential]), Array{Int}(undef, length(parents)), parents),
```
Comment (mykelk) [inline src/CPDs/categorical_cpd.jl:122]: should be `p`

File: src/CPDs/conditional_linear_gaussian_cpd.jl [hunk 4]
```diff
@@ -70,8 +70,8 @@ function Distributions.fit(::Type{ConditionalLinearGaussianCPD},
     # ---------------------
     # identify discrete and continuous parents
 
-    parents_disc = filter(p->eltype(data[!,p]) <: Int, parents)
-    parents_cont = filter(p->eltype(data[!,p]) <: AbstractFloat, parents)
+    parents_disc = filter(potential->eltype(data[!,potential]) <: Int, parents)
```
Comment (mykelk) [inline src/CPDs/conditional_linear_gaussian_cpd.jl:73]: should be `p`

File: src/CPDs/conditional_linear_gaussian_cpd.jl [hunk 5]
```diff
@@ -70,8 +70,8 @@ function Distributions.fit(::Type{ConditionalLinearGaussianCPD},
...
+    parents_cont = filter(potential->eltype(data[!,potential]) <: AbstractFloat, parents)
```
Comment (mykelk) [inline src/CPDs/conditional_linear_gaussian_cpd.jl:74]: should be `p`

File: src/CPDs/conditional_linear_gaussian_cpd.jl [hunk 6]
```diff
@@ -85,8 +85,8 @@ function Distributions.fit(::Type{ConditionalLinearGaussianCPD},
 
         parental_ncategories = Array{Int}(undef, nparents_disc)
         dims = Array{UnitRange{Int64}}(undef, nparents_disc)
-        for (i,p) in enumerate(parents_disc)
-            parental_ncategories[i] = infer_number_of_instantiations(data[!,p])
+        for (i,potential) in enumerate(parents_disc)
```
Comment (mykelk) [inline src/CPDs/conditional_linear_gaussian_cpd.jl:88]: should be `p`

File: src/CPDs/conditional_linear_gaussian_cpd.jl [hunk 7]
```diff
@@ -85,8 +85,8 @@ function Distributions.fit(::Type{ConditionalLinearGaussianCPD},
...
+            parental_ncategories[i] = infer_number_of_instantiations(data[!,potential])
```
Comment (mykelk) [inline src/CPDs/conditional_linear_gaussian_cpd.jl:89]: same here

File: src/Factors/factors_main.jl [hunk 8]
```diff
@@ -5,6 +5,12 @@
 # THE MOST BASIC ASSUMPTION IS THAT ALL VARIABLES ARE CATEGORICAL AND THEREFORE
 # Base.OneTo WORTHY. IF THAT IS VIOLATED, NOTHING WILL WORK
 
+nodeconvert(::Type{NodeNames}, names::NodeNameUnion) = names
+
```
Comment (mykelk) [inline src/Factors/factors_main.jl:9]: drop unnecessary blank lines

File: test/test_cpds.jl [hunk 9]
```diff
@@ -193,8 +193,8 @@ end
 let
     a = StaticCPD(:a, Bernoulli(0.5))
     b = StaticCPD(:b, Bernoulli(0.6))
-    p = [:a,:b]
-    c = ParentFunctionalCPD{Bernoulli}(:c, p, (seq,par)->begin
+    potential = [:a,:b]
```
Comment (mykelk) [inline test/test_cpds.jl:196]: should be `p`

**Review by mykelk — CHANGES_REQUESTED** [review]

**Pull request merged**